### PR TITLE
Fix `broadcast_in_dim` support in NVFuser Frontend

### PR DESCRIPTION
**Pull request by kevinstephano**
This PR primarily addresses augmenting the frontend to properly support `broadcast_in_dim`.  This required make a new version of the `define_tensor()` that takes in the `size` and `strides` of input tensors in order to properly determine broadcasts.

This PR also has a fix for the `python_example.py` that broke when a new argument was added to reductions to allow the user to specify an output Data Type.

`define_tensor()` Interface Example:

```
fusion2 = Fusion()

input1 = torch.ones(1, 1, 4, device='cuda')
input2 = torch.ones(2, 3, 4, device='cuda')

with FusionDefinition(fusion2) as fd :
    t0 = fd.define_tensor(sizes=input1.size(), strides=input1.stride())
    t1 = fd.define_tensor(sizes=input2.size(), strides=input2.stride())

    fd.add_input(t0)
    fd.add_input(t1)

    t0_b = fd.Ops.broadcast_in_dim(t0, [2, 3, 4], [0, 1, 2])
    print("Broadcast TensorView", t0_b)
    t2 = fd.Ops.add(t0_b, t1)

    fd.add_output(t2)
```
Print statement of defined broadcast tensor:

```
Broadcast TensorView T2_l[ sbS6{1}, sbS7{1}, iS8{i2} ] DataType: float Contiguity: ttt
```

### CODE REVIEW

**Review by mruberry — COMMENTED** [review]
Can this also update the existing test in test_prims.py to use nvFuser as a trace executor? 

https://github.com/pytorch/pytorch/blob/dfde877c0b1c0e5d1ef7b28082186cf019a4af57/test/test_prims.py#L27

As we spoke about briefly offline, maybe that test needs to do something after the call to broadcast_in_dim (like an addition)?

**Comment by kevinstephano** [comment]
I had to update the prims executor to use the new `define_tensor()` flavor and I also updated the prims testing of `broadcast_in_dim` to include an additional `prims::add` instruction of zero tensor to realize the broadcast with nvfuser enabled.

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -187,46 +192,76 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+              }
+            }
+
+            std::vector<bool> contig_info(strides.size(), false);
+            for (int i = contig_info.size() - 1; i >= 0; --i) {
+              if (i == static_cast<int>(contig_info.size() - 1)) {
+                contig_info[i] = (strides[i] == 1);
+              } else {
+                contig_info[i] =
+                    (strides[i] == (strides[i - 1] * sizes[i - 1]));
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:228]: Caught you again here~

Per the line above, higher dimension should be the inner dimension, so here `i-1` should be `i+1` instead.

Comment (mruberry) [reply torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:228]: Feel free to add test cases to the test in test_prims.py, too!

Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:228]: Right now, the correct contiguity logic is causing a failure in this test that I am trying to track down.

```
            # Adds outermost, expands, and unsqueezes
            a = make_arg((1, 2, 3))
            b = make_arg((4, 1, 7, 2, 3, 3), low=0.0, high=0.0)
            result = fn(a, b, (1, 3, 4))

            self.assertEqual(result.shape, b.shape)
            a.unsqueeze_(3)
            a.unsqueeze_(1)
            a.unsqueeze_(0)
            self.assertEqual(a.expand_as(result), result)
```

Specifically the comparison of `a.expand_as(result)` is failing.

**Comment by kevinstephano** [comment]
I fixed the contiguity indexing logic.  I believe the last `broadcast_in_dim` test is exposing an nvfuser indexing with contiguity.

**Comment by kevinstephano** [comment]
This PR is ready to go.  The bug that I found lead to currently having to comment out a test case for `broadcast_in_dim` was fixed in NVFuser's devel branch.  So, once the pointwise scheduler changes propagate upstream, this bug will go away.  The Python Frontend code, itself, is healthy and ready to go.

File: test/test_prims.py [hunk 2]
```diff
@@ -18,63 +18,68 @@ class TestPrims(TestCase):
...
-        # TODO: FIXME:
-        # for executor in ('aten', 'nvfuser'):
-        for executor in ("aten",):
+        for executor in ('aten', 'nvfuser'):
             fn = partial(traced, executor=executor)
             # Same shape
             shape = (5, 5)
             a = make_arg(shape)
-            result = fn(a, shape, (0, 1))
+            b = make_arg(shape, low=0.0, high=0.0)
```
Comment (mruberry) [inline test/test_prims.py:35]: Why make b a tensor of just zeros?

**Review by mruberry — APPROVED** [review]
Stamped!

File: test/test_prims.py (hunk 2 above)
Comment (jjsjann123) [reply test/test_prims.py:35]: Same question here, wasn't this going to hide the indexing issue into b (if there's any).

Is this just to make test easier?

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp [hunk 3]
```diff
@@ -187,46 +192,76 @@ void initNvFuserPythonBindings(PyObject* module) {
...
-          py::arg("contiguity"),
+          py::arg("sizes"),
+          py::arg("strides"),
           py::arg("dtype") = torch::jit::fuser::cuda::DataType::Float,
           py::return_value_policy::reference)
-      .def_static(
+      .def(
           "define_constant",
-          [](double val) -> torch::jit::fuser::cuda::Val* {
+          [](FusionDefinitionContextManager& self,
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:242]: Does this change imply that `define_constant` & `define_scalar` weren't functional? I don't see we added tests for those in this PR neither :stuck_out_tongue_winking_eye:

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp [hunk 4]
```diff
@@ -187,46 +192,76 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+                    self.fusionPtr()->zeroVal(), IrBuilder::create<Int>()));
+              }
+            }
+
+            std::vector<bool> contig_info(strides.size(), false);
+            for (int i = contig_info.size() - 1; i >= 0; --i) {
+              if (i == static_cast<int>(contig_info.size() - 1)) {
+                contig_info[i] = (strides[i] == 1);
+              } else {
+                contig_info[i] =
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:227]: I think we are all good here.

I'm a little uncertain how collapsing across a broadcasted dimension would be handled by codegen. But that's a separate topic we might want to investigate separately.

i.e. -> tensor with size [4, 1, 4] & stride [8, 8, 1]
I'll try that on our devel branch

**Review by jjsjann123 — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp (hunk 4 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:227]: NVM.. double checked the kernel and everything looks fine.

Turns out we are not doing anything special for size-1 broadcasting, in terms of indexing... So collapsing does stop properly there.

File: test/test_prims.py (hunk 2 above)
Comment (kevinstephano) [reply test/test_prims.py:35]: I made the `b` Tensor full of zeros because I am really testing the `broadcast_in_dim` operation but I need an `add` or any other operation after the `broadcast_in_dim` to realize the broadcasted output.  Using zeros made it easier to figure out whether the broadcast was successful in moving the values to the right spots without polluting them non-zero value.

File: torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp (hunk 3 above)
Comment (kevinstephano) [reply torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp:242]: `define_constant` and `define_scalar` were perviously added and they are functional in the sense that they can do something.  Tests just haven't been added in the `prims` world.
